cfme/tests/cloud_infra_common/test_provisioning.py

```diff
@@ -389,38 +389,60 @@ def test_cloud_provision_from_template_with_attached_disks(
-            device_mapping.append((volume, device_name.format(chr(ord("b") + i))))
-        # Set up automate
+    volumes = provider.mgmt.volume_configurations(1, n=disks)
 
-        method = modified_request_class.methods.instantiate(name="openstack_PreProvision")
+    @request.addfinalizer
+    def delete_volumes():
+        for volume in volumes:
+            provider.mgmt.delete_volume(volume)
+
+    # Set up automate
+    for i, volume in enumerate(volumes, 0):
+        if i == 0:
+            provider.mgmt.capi.volumes.set_bootable(volume, True)
+            device_mapping.append(
+                {'boot_index': 0,
+                'uuid': volume,
+                'device_name': device_name.format(chr(ord("a") + i))})
+        else:
+            device_mapping.append(
+                {'boot_index': -1,
+                'uuid': volume,
+                'device_name': device_name.format(chr(ord("a") + i))})
+
+    method = modified_request_class.methods.instantiate(name="openstack_PreProvision")
+
+    view = navigate_to(method, 'Details')
+    former_method_script = view.script.get_value()
+    with update(method):
+        disk_mapping = []
+        for mapping in device_mapping:
+            disk_mapping.append(ONE_FIELD.format(**mapping))
+        method.script = VOLUME_METHOD.format(",\n".join(disk_mapping))
 
+    @request.addfinalizer
+    def _finish_method():
         with update(method):
-            disk_mapping = []
-            for mapping in device_mapping:
-                disk_mapping.append(ONE_FIELD.format(*mapping))
-            method.script = VOLUME_METHOD.format(", ".join(disk_mapping))
-
-        def _finish_method():
-            with update(method):
-                method.script = """prov = $evm.root["miq_provision"]"""
-        request.addfinalizer(_finish_method)
-
-        instance = appliance.collections.cloud_instances.create(vm_name,
-                                                                provider,
-                                                                form_values=inst_args)
-
-        for volume_id in volumes:
-            soft_assert(vm_name in provider.mgmt.volume_attachments(volume_id))
-        for volume, device in device_mapping:
-            soft_assert(provider.mgmt.volume_attachments(volume)[vm_name] == device)
-        instance.mgmt.delete()  # To make it possible to delete the volume
+            method.script = former_method_script
+
+    instance = appliance.collections.cloud_instances.create(vm_name,
+                                                            provider,
+                                                            form_values=inst_args)
+
+    @request.addfinalizer
+    def delete_vm_and_wait_for_gone():
+        instance.mgmt.delete()
         wait_for(lambda: not instance.exists_on_provider, num_sec=180, delay=5)
 
+    for volume_id in volumes:
+        assert vm_name in provider.mgmt.volume_attachments(volume_id)
```

Comment:
Sure you want to change this from a `soft_assert`? It allows us to get way more feedback in the test if the volumes are not lining up. With only one failing and killing the test, we get very little info.

With the soft_assert, we can maybe see that the volume ID's are all jumbled, or all empty, or whatever the case may be. We also get information about if its just the volumes, or if the following device_mapping assertions fail as well.

Reply:
@mshriver:
```
def test_soft_assert(soft_assert):
    for a in [False, "fff", True]:
        soft_assert(a).
```
```
____________________________________________ test_soft_assert ____________________________________________

item = <Function 'test_soft_assert'>

    @pytest.mark.hookwrapper
    def pytest_runtest_call(item):
        """pytest hook to handle :py:func:`soft_assert` fixture usage"""
        yield
        if 'soft_assert' in item.fixturenames and _thread_locals.caught_asserts:
>           raise SoftAssertionError(_thread_locals.caught_asserts)
E           SoftAssertionError: 
E           soft_assert(a) (cfme/tests/test_soft_assert.py:3)

```
Doesn't really help much.

Reply:
Understood, but I think you may have been missing the soft assert exception context in the results where soft assert fails. This provides as rich of a message as you want along with the soft_assert. It should be passed as a second option to the  method, and can include whatever information you want so that its clear what failed.

Example of proper use of soft_assert with context:
```
# this statement is in a for loop, where target is the iterated value
soft_assert(False, '0 occurrence of {} found on the timeline of {}'.format(
                        self.event, target))```

And the resulting traceback:
```E           0 occurrence of policy found on the timeline of Host(name=<redacted>, provider=VMwareProvider(<redacted>), hostname=None, ip_address=None, custom_ident=None, host_platform=None, ipmi_address=None, mac_address=None, credentials={}, ipmi_credentials=None, interface_type='lan') (cfme/tests/infrastructure/test_timelines.py:300)
E           0 occurrence of policy found on the timeline of VMwareProvider(<redacted>) (cfme/tests/infrastructure/test_timelines.py:300)

cfme/fixtures/soft_assert.py:64: SoftAssertionError
SoftAssertionError

0 occurrence of policy found on the timeline of Cluster(name=u'Cluster', provider=VMwareProvider(<redacted>) (cfme/tests/infrastructure/test_timelines.py:300)
0 occurrence of policy found on the timeline of Host(name=<redacted>, provider=VMwareProvider(<redacted>), hostname=None, ip_address=None, custom_ident=None, host_platform=None, ipmi_address=None, mac_address=None, credentials={}, ipmi_credentials=None, interface_type='lan') (cfme/tests/infrastructure/test_timelines.py:300)
0 occurrence of policy found on the timeline of VMwareProvider(<redacted>) (cfme/tests/infrastructure/tes...```


As far as debugging goes, I'm not inclined to make test coverage concessions in order to facilitate debugging. We can easily make local changes to facilitate automatic triggering of a debugger. In this case you could have set a break-point at the soft_assert call, and checked the state of the values being asserted before the call is made, instead of relying on automatic triggering.

We're trading valuable assertion data at the time of test for the possibility of triggering a PDB instance when running locally. If we immediately fail on the first assert here, then we miss asserting any remaining volumes that haven't been asserted against yet, and all of the device_mapping assertion below.  We would end up with a failed test in automation, with less information about what was wrong (or right) with the volumes and devices.